src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -0,0 +1,348 @@
+        }
+
+        private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBufferScope, ICodeCleanUpExecutionContext context, CancellationToken cancellationToken)
+        {
+            using (var cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(context.OperationContext.UserCancellationToken, cancellationToken))
+            {
+                cancellationToken = cancellationTokenSource.Token;
+
+                var buffer = textBufferScope.SubjectBuffer;
+                if (buffer != null)
+                {
+                    using (var scope = context.OperationContext.AddScope(allowCancellation: true, description: EditorFeaturesResources.Applying_changes))
+                    {
+                        var progressTracker = new ProgressTracker((description, completed, total) =>
+                        {
+                            if (scope != null)
+                            {
+                                scope.Description = description;
+                                scope.Progress.Report(new ProgressInfo(completed, total));
+                            }
+                        });
+
+                        var document = buffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                        var codeCleanupService = document.GetLanguageService<ICodeCleanupService>();
+
+                        // TODO: enable all diagnostics for now, need to be replace by inclusion/ exclusion list from .editorconfig
+                        // https://github.com/dotnet/roslyn/issues/30163
+                        var allDiagnostics = codeCleanupService.GetAllDiagnostics();
+
+                        var enabedDiagnosticSets = ArrayBuilder<DiagnosticSet>.GetInstance();
+
+                        foreach (var diagnostic in allDiagnostics.Diagnostics)
+                        {
+                            foreach (var diagnosticId in diagnostic.DiagnosticIds)
+                            {
+                                if (context.EnabledFixIds.IsFixIdEnabled(diagnosticId))
+                                {
+                                    enabedDiagnosticSets.Add(diagnostic);
+                                    break;
+                                }
+                            }
+                        }
+
+                        var isRemoveUnusedUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(RemoveUnusedImportsFixId);
+                        var isSortUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(SortImportsFixId);
+                        var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
+                            new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+
+                        var newDoc = await codeCleanupService.CleanupAsync(
+                            document, enabledDiagnostics, progressTracker, cancellationToken);
```

Comment:
possibly document that it's intentional to not have ConfigureAwait?  I'm guessing we want to come back to the calling thread for the TryApply call below?  But it's unclear if this is intentional or not.

Reply:
Added `ConfigureAwait(true)` and an explicit switch to the main thread.